llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -1145,31 +1169,71 @@ struct AAAMDWavesPerEU : public AAAMDSizeRangeAttribute {
     auto &InfoCache = static_cast<AMDGPUInformationCache &>(A.getInfoCache());
     ChangeStatus Change = ChangeStatus::UNCHANGED;
 
+    Function *F = getAssociatedFunction();
+
+    const auto *AAFlatWorkGroupSize = A.getAAFor<AAAMDFlatWorkGroupSize>(
+        *this, IRPosition::function(*F), DepClassTy::REQUIRED);
+    if (!AAFlatWorkGroupSize || !AAFlatWorkGroupSize->isValidState()) {
+      LLVM_DEBUG(
+          dbgs() << '[' << getName()
+                 << "] AAAMDFlatWorkGroupSize is unavailable or invalid.\n");
+      return ChangeStatus::UNCHANGED;
+    }
+
+    if (AAFlatWorkGroupSize->isAtInitialState()) {
+      LLVM_DEBUG(dbgs() << '[' << getName()
+                        << "] AAAMDFlatWorkGroupSize is still at initial "
+                           "state. Skip the update.\n");
+      return ChangeStatus::UNCHANGED;
+    }
```

Comment:
The isAtInitialState check is a superset of the above isValidState, just replace the isValidState call?

Reply:
An initial state is always a valid state, but the reverse is not necessarily true. Once the value is updated, it is no longer in the initial state; however, it remains valid. As such, updates should continue, and the manifest function must still be invoked.

Reply:
There are four states to consider: valid, invalid, best, and worst. An AA typically begins in the _best_ state and transitions toward a "worse" state until reaching the _worst_ state. For most AAs, the invalid state is the worst state, but this is not always true, especially for the `IntegerRangeState`, which we use extensively here.

In the case of `IntegerRangeState`, the worst state is considered invalid **only if** `Known` is set to the worst value (`~0U` by default). Once `Known` is modified, there is no longer invalid state, meaning that `manifest` will always run, and half-baked values are likely to be added. This is why I avoided updating `Known` in #113018 and all related PRs in this stack unless the changes were definitive (i.e., all call sites were seen).

The situation becomes more complex because the _best_ state does not actually represent a valid pair of values in terms of the attribute. However, it must still be considered valid so that updates can occur, allowing `Assumed` to be updated accordingly.

Complications arise further because this AA relies on another AA that uses range state. Similarly, the _best_ state of the dependent AA is not a valid pair of values either. If the dependent AA remains in its _best_ (or _initial_, as introduced in this PR) state, we must skip the update, as its values are effectively nonsensical.